File: NeetCode 150/Backtracking/131. Palindrome Partitioning.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        res = []
        subset = []

        def isPalindrome(l,r):
            while (l<r):
                if s[l] != s[r]:
                    return False
                l += 1
                r -= 1
            return True
        
        def backtrack(start):
             # If we've used the entire string, record the partition
            if start == len(s):
                res.append(subset[:]) # make a copy
                return
             # Try all possible end positions
            for end in range(start,len(s)):
                if isPalindrome(start,end):
                    subset.append(s[start:end+1])
                    backtrack(end+1)
                    subset.pop()

        backtrack(0)
        return res

File: NeedCode 150 (JAVA)/Backtracking/131. Palindrome Partitioning.java
class Solution {
    private List<List<String>> res = new ArrayList<>();
    private List<String> subset = new ArrayList<>();

    public List<List<String>> partition(String s) {
        backtrack(0,s);
        return res;
    }

    //start is the index where we need to partition next.

    private void backtrack(int start, String s){
        if(start == s.length()) 
            res.add(new ArrayList<>(subset));
        
        //This loop tries every possible substring starting at start.
        for(int end = start;end<s.length();end++){
            if(isPalindrome(s,start,end)){
                subset.add(s.substring(start,end+1)); // ecause substring excludes the end index.
                backtrack(end + 1, s);
                subset.remove(subset.size() - 1);
            }
        }
    }

    private boolean isPalindrome(String s, int l, int r){
        while(l<r){
            if(s.charAt(l) != s.charAt(r))
                return false;
            l+=1;
            r-=1;
        }

        return true;
    }

}

//The only difference is what you choose:

//Subsets: choose an element.
//Combination Sum: choose a number.
//Palindrome Partitioning: choose a palindromic substring.
